pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A transform for decomposing quantum circuits into user defined gate sets. Offers an alternative to the more device-focused decompose transform.
+"""
+# pylint: disable=protected-access, too-many-arguments
+
+import warnings
+from collections.abc import Callable, Generator, Sequence
+from typing import Optional
+
+import pennylane as qml
+from pennylane.tape import QuantumScript
+from pennylane.transforms.core import transform
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+def _operator_decomposition_gen(
+    op: qml.operation.Operator,
+    acceptance_function: Callable[[qml.operation.Operator], bool],
+    decomposer: Callable[[qml.operation.Operator], Sequence[qml.operation.Operator]],
+    max_expansion: Optional[int] = None,
+    current_depth=0,
+) -> Generator[qml.operation.Operator, None, None]:
+    """A generator that yields the next operation that is accepted."""
+    max_depth_reached = False
+    if max_expansion is not None and max_expansion <= current_depth:
+        max_depth_reached = True
+    if acceptance_function(op) or max_depth_reached:
+        yield op
+    else:
+        try:
+            decomp = decomposer(op)
+            current_depth += 1
+        except qml.operation.DecompositionUndefinedError as e:
+            raise UserWarning(
+                f"Operator {op.name} has no supported decomposition and is not in the gate set."
+            ) from e
+
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(
+                sub_op,
+                acceptance_function,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                current_depth=current_depth,
+            )
+
+
+@transform
+def apply_decomposition(tape, gate_set=None, gate_rules=None, max_expansion=None):
```

Comment:
`apply_decomposition` feels a bit redundant, as it has the same meaning as just `decompose` (which is shorter).

I noticed the original shortcut story requirements had this named `decompose`, was there a reason for renaming it? (cc @lillian542 as this might have happened during code review?)

Reply:
@josh146, I believe we're avoiding `decompose` as there is already the device focused function (`qml.devices.preprocess.decompose`) with the same name. Either we deprecate the name of the existing `decompose` function and transfer it to this function or come up with a new name for this 😄.

Reply:
This one would take precedence for users, so I'd be ok keeping the `decompose` name for this one, and then changing the name of the more developer-focused one. Given they are in different namespaces, they should be able to have the same name for a time. 

There shouldn't be any software reason to keep different names, so we can temporarily have them overlap.